### DISCUSSION TITLE
How does JS events execute in Blazor?

### DISCUSSION THREAD

**Discussion by zxyao145** [dotnet/aspnetcore#52365] [Q&A] [answered]
For the following code, JS will output "1 child mouseup", "2 child mouseup end", "3 parent mouseup" in sequence

```html
<div id="parent">
    <div id="child">
      child
    </div>
  </div>
  <script>
    document.getElementById("parent").addEventListener("mouseup", () => {
      console.log("3 parent mouseup");
    });

    document.getElementById("child").addEventListener("mouseup", () => {
      console.log("1 child mouseup");
      // Simulate a time-consuming operation
      for (let i = 0; i < 10_0000; i++) {
        var text = i + "j: ";
        for (let j = 0; j < 10_000; j++) {
          text += " " + j;
        }
      }
      console.log("2 child mouseup end");
    });
  </script>
```
![image](https://github.com/dotnet/aspnetcore/assets/24973084/7635f9e3-b7d0-4756-a1f1-bef561b224a6)

If we use Blazor to implement it, it will output "1 child mouseup", "3 parent mouseup", "2 child mouseup end" in sequence

```razor
<div id="parent"
     @onmouseup="OnParentMouseUp">
    <div id="child"
         @onmouseup="OnChildMouseUp">
        child
    </div>
</div>


@code {

    private void OnParentMouseUp()
    {
        Console.WriteLine("3 Parent mouseup");
    }

    private async Task OnChildMouseUp()
    {
        Console.WriteLine("1 Child mouseup");

        await Task.Delay(10000);

        Console.WriteLine("2 Child mouseup end");
    }
}
```

![image](https://github.com/dotnet/aspnetcore/assets/24973084/b41a7982-3e92-4aa8-adaa-94c5dd61a692)

**Comment by En3Tho** [comment]
Your js code is simulating load synchronously, it doesn't yield to the thread pool / event loop while your blazor code does.

**Reply by zxyao145** [reply]
Yes, how does Blazor handle events asynchronously, does it use `setTimeout`? I searched for the source code, but I'm sorry I couldn't find it. Can you help me point it out?

**Reply by En3Tho** [reply, marked as answer]
Sorry, I can't point you to the code.  I just pointed that if you did simulate load synchronously in blazor like you did in js output would've been the same I think.

I think you should check out dotnet/runtime repo sources. Blazor runs on Mono runtime. That's there Mono's sources are, not here. Or maybe someone from the runtime team can answer here.

I belive @pavelsavara can help with this question

**Reply by zxyao145** [reply]
Get it, thank you very much for your answer!

**Reply by pavelsavara** [reply]
https://github.com/dotnet/runtime/blob/ddd663d257aa2206583287309a2109e1835f1b2f/src/mono/wasm/runtime/scheduling.ts#L66

**Reply by En3Tho** [reply]
@pavelsavara Thank you!

**Comment by zxyao145** [comment]
> https://github.com/dotnet/runtime/blob/ddd663d257aa2206583287309a2109e1835f1b2f/src/mono/wasm/runtime/scheduling.ts#L66

@pavelsavara Thank you very much for your answer! This is exactly the source code I want to read.

**Comment by zxyao145** [comment]
@pavelsavara Hi, I have another question to ask. 

I have two components: the child component uses `@onclick` to add event, while the parent component uses native JS to add events through` IJSRuntime`. However, using `@onclick:stopPromotion="@true"` on the child component does not prevent the parent component from responding to click events.

Is Blazor's handling of events similar to React?

Child.razor
```razor
<span style="border: 1px solid red;"
      @onclick="OnClick"
      @onclick:stopPropagation="@true">
    child
</span>

@code {
    private void OnClick()
    {
        Console.WriteLine("child click");
    }
}
```


Parent.razor
```razor
@inject IJSRuntime JsRuntime

<div @ref="@_ref" id="parent">
    <Child />
</div>

@code
{

    private ElementReference _ref;
    protected override Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            JsRuntime.InvokeVoidAsync("addClick", _ref);
        }
        return base.OnAfterRenderAsync(firstRender);
    }
}
```

`addClick` method:
```js
window.addClick = function (dom) {
    dom.addEventListener("click", (e) => {
        console.log("on click", dom.getAttribute("id"));
    })
}
```

the log:
![image](https://github.com/dotnet/aspnetcore/assets/24973084/9937730d-c416-49cc-a699-a8e8284c6f1f)

**Reply by pavelsavara** [reply]
I don't know. @javiercn  ?

**Reply by zxyao145** [reply]
@javiercn Hi, could you help me take a look at this question?
